### PR TITLE
[Runtime] Fix incorrect memory ordering in ConcurrentReadableArray/HashMap

### PR DESCRIPTION
**Pull request by mikeash**
When reallocating storage, the storing the pointer to the new storage had insufficiently strong ordering. This could cause writers to check the reader count before storing the new storage pointer. If a reader then came in between that load and store, it would end up using the old storage pointer, while the writer could end up freeing it.

Also adjust the Concurrent.cpp tests to test with a variety of reader and writer counts. Counterintuitively, when freeing garbage is gated on there being zero readers, having a single reader will shake out problems that having lots of readers will not. We were testing with lots of readers in order to stress the code as much as possible, but this resulted in it being extremely rare for writers to ever see zero active readers.

rdar://69798617

### CODE REVIEW

**Comment by mikeash** [comment]
@swift-ci please test

**Comment by mikeash** [comment]
@swift-ci please test windows platform

File: include/swift/Runtime/Concurrent.h [hunk 1]
```diff
@@ -527,7 +527,7 @@ template <class ElemTy> struct ConcurrentReadableArray {
       
       storage = newStorage;
       Capacity = newCapacity;
-      Elements.store(storage, std::memory_order_release);
+      Elements.store(storage, std::memory_order_seq_cst);
```
Comment (davezarzycki) [inline include/swift/Runtime/Concurrent.h:530]: I think the fix is correct but the reasoning doesn't feel right to me at all. The problem isn't that the readers are picking up the old value but that the writer isn't reasoning correctly about whether there are readers or not. After all, this proposed fix is on the writer side of the design.

On x86, the only C++11 memory model hint that matters "sequential consistency" and even then, it only matters for stores. There are two reasons for this: 1) atomic x86 instructions don't support anything other than sequential consistency and 2) the only reordering allowed by x86's "total store order" memory model is newer loads being moved ahead of older stores.

By switching `Elements.store(...)` to sequential consistency, we're forcing `ReaderCount.load()` to happen after the store.

I think if one is going to use `std::memory_order_seq_cst`, then one really ought to document at least one example in a comment of an actual subsequent load in the same thread that must come after the given store.

**Review by davezarzycki — APPROVED** [review]
For whatever it may be worth, I think this bug would have been avoided if `push_back` were made fully concurrent instead of using a mutex. That is easy to say of course because that kind of redesign would be a lot of work.

File: include/swift/Runtime/Concurrent.h (hunk 1 above)
Comment (davezarzycki) [reply include/swift/Runtime/Concurrent.h:530]: Here is the ASCII art of what I think is happening:
```
Effective program order due to out-of-order loads.

Writer                 | Reader
-----------------------+--------------------
............ ReaderCount == 0 ..............
                       |
 -> ReaderCount.load() |
/                      |
|                      | incrementReaders()
|                      | storage = Elements.load(SWIFT_MEMORY_ORDER_CONSUME);
|                      |
| Elements.store()     |
\                      |
 *ReaderCount.load()   |

Now the writer wrongly thinks there are zero readers of the old data.
```

**Comment by mikeash** [comment]
@swift-ci please test

**Pull request merged**